sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala

```diff
@@ -0,0 +1,109 @@
+package sigmastate.interpreter
+
+import java.math.BigInteger
+
+import sigmastate.UncheckedTree
+import sigmastate.Values.SigmaBoolean
+import sigmastate.basics.FirstProverMessage
+import sigmastate.basics.VerifierMessage.Challenge
+
+/**
+  * A hint for a prover which helps the prover to prove a statement. For example, if the statement is "pk1 && pk2",
+  * and the prover knows only a secret for the public key pk1, the prover fails on proving without a hint. But if the
+  * prover knows that pk2 is known to another party, the prover may prove the statement (with an empty proof for "pk2").
+  */
+trait Hint
+
+/**
+  * A hint which is indicating that a secret associated with its public image "image" is already proven.
+  */
+trait SecretProven extends Hint {
+  val image: SigmaBoolean
+  val challenge: Challenge
+  val uncheckedTree: UncheckedTree
+}
+
+/**
+  * A hint which contains a proof-of-knowledge for a secret associated with its public image "image".
+  */
+case class RealSecretProof(image: SigmaBoolean,
+                           challenge: Challenge,
+                           uncheckedTree: UncheckedTree) extends SecretProven
+
+/**
+  * A hint which contains a proof-of-knowledge for a secret associated with its public image "image".
+  */
+case class SimulatedSecretProof(image: SigmaBoolean,
+                                challenge: Challenge,
+                                uncheckedTree: UncheckedTree) extends SecretProven
+
+
+/**
+  * A family of hints which are about a correspondence between a public image of a secret image and prover's commitment
+  * to randomness ("a" in a sigma protocol).
+  */
+trait CommitmentHint extends Hint {
+  val image: SigmaBoolean
+  val commitment: FirstProverMessage
+}
+
+/**
+  * A hint which a commitment to randomness associated with a public image of a secret, as well as randomness itself.
+  * Please note that this randomness should be kept in secret by the prover.
+  *
+  * @param image      - image of a secret
+  * @param randomness - randomness
+  * @param commitment - commitment to randomness used while proving knowledge of the secret
+  */
+case class OwnCommitment(override val image: SigmaBoolean, randomness: BigInteger, commitment: FirstProverMessage) extends CommitmentHint
+
+/**
+  * A hint which contains a commitment to randomness associated with a public image of a secret.
+  *
+  * @param image      - image of a secret
+  * @param commitment - commitment to randomness used while proving knowledge of the secret
+  */
+case class RealCommitment(override val image: SigmaBoolean, commitment: FirstProverMessage) extends CommitmentHint
```

Comment:
Own commitment is also a real commitment (as opposed to simulated commitment)
Its a bit confusing to outsiders who don't know the actual protocol. Perhaps, its better to have the hierarchy as:

`RealCommitment` and `SimulatedCommitment` as subtypes of `CommitmentHint`

and then `OwnCommitment` as subtype of `RealCommitment`

Reply:
We can't inherit a case class from a case class. Then we can have RealCommitment abstract entity and descendants RealOtherCommitment and RealOwnCommitment, but I'm not sure if it is better.